servers/src/main/java/tachyon/master/file/FileSystemMaster.java

```diff
@@ -427,16 +456,43 @@ public void completeFile(long fileId) throws FileDoesNotExistException, BlockInf
     }
   }
 
-  void completeFileInternal(List<Long> blockIds, long fileId, long fileLength,
-      long opTimeMs) throws FileDoesNotExistException {
+  void completeFileInternal(List<Long> blockIds, long fileId, long fileLength, long opTimeMs)
+      throws FileDoesNotExistException, InvalidPathException {
+    // This function should only be called from within synchronized (mInodeTree) blocks.
     mDependencyMap.addFileCheckpoint(fileId);
     InodeFile inodeFile = (InodeFile) mInodeTree.getInodeById(fileId);
     inodeFile.setBlockIds(blockIds);
-    inodeFile.setComplete(fileLength);
+    inodeFile.setCompleted(fileLength);
     inodeFile.setLastModificationTimeMs(opTimeMs);
+    // Mark all parent directories that have a UFS counterpart as persisted.
+    propagatePersisted(inodeFile);
   }
 
-  private void completeFileFromEntry(CompleteFileEntry entry) {
+  /**
+   * Propagates the persisted status to all parents of the given inode in the same mount partition.
+   *
+   * @param inode the inode to start the propagation at
+   * @throws FileDoesNotExistException if a non-existent file is encountered
+   * @throws InvalidPathException if an invalid path is encountered
+   */
+  private void propagatePersisted(Inode inode) throws FileDoesNotExistException,
+      InvalidPathException {
+    if (!inode.isPersisted()) {
+      return;
+    }
+    Inode handle = inode;
+    while (handle.getParentId() != InodeTree.NO_PARENT) {
+      handle = mInodeTree.getInodeById(handle.getParentId());
+      handle.setPersisted(true);
```

Comment:
Well, this is a bug. The operation should be journaled, while the existing journaling of the event in journalCreatePathResult() should be removed. I have modified the code accordingly.